Sources/CartonCLI/Commands/Init.swift

```diff
@@ -25,43 +25,85 @@ struct Init: ParsableCommand {
+  @Option(
+    name: .long,
+    help: """
+          The name of the project, if it's different from the current working directory's name.
+          If this option is set, carton will create the package for the project in a subdirectory by the given name.
+          Otherwise, the package will be created in the current working directory, with the package named after the directory's.
+          """
+  )
+  var name: String?
 
   func run() throws {
     let terminal = InteractiveWriter.stdout
 
-    guard let name = name ?? localFileSystem.currentWorkingDirectory?.basename else {
-      terminal.write("Project name could not be inferred\n", inColor: .red)
-      return
-    }
-    guard let currentDir = localFileSystem.currentWorkingDirectory else {
-      terminal.write("Failed to get current working directory.\n", inColor: .red)
-      return
+    guard let currentDirectory = localFileSystem.currentWorkingDirectory else {
+      throw ProjectInitialisationError.cannotFindCurrentDirectory
     }
+
+    /// The kind of template the project is to be created with.
+    /// - Note: This value is of the type `Templates`, not `Template`.
     let template = self.template ?? .basic
+
+    let projectName = name ?? currentDirectory.basename
+    let projectPath = name == nil ? currentDirectory : currentDirectory.appending(component: projectName)
+    
+    // FIXME: Replace with `NSString.abbreviatingWithTildeInPath`?
+    let packagePathRelativeToHomeDirectory = projectPath.relative(to: localFileSystem.homeDirectory)
+
     terminal.write("Creating new project with template ")
     terminal.write("\(template.rawValue)", inColor: .green)
     terminal.write(" in ")
-    terminal.write("\(name)\n", inColor: .cyan)
+    terminal.write("~/\(packagePathRelativeToHomeDirectory.pathString)\n", inColor: .cyan)
 
-    guard let packagePath = self.name == nil ?
-      localFileSystem.currentWorkingDirectory :
-      AbsolutePath(name, relativeTo: currentDir)
-    else {
-      terminal.write("Path to project could be created.\n", inColor: .red)
-      return
+    do {
+      // `LocalFileSystem.createDirectory` wraps `FileManager.createDirectory`, which throws an NSError.
+      try localFileSystem.createDirectory(projectPath)
+    } catch let nsError as NSError {
+        throw ProjectInitialisationError.cannotCreateDirectory(path: packagePathRelativeToHomeDirectory, reason: nsError.localizedDescription)
+    }
+    
+    // FIXME: Handle this error-catching in each `Template.create()` instead.
+    do {
+      // FIXME:  The first `template` is of type `Templates`, and two consecutive `template` looks odd.
+      // Maybe rename `Templates` as `TemplateKind`, and refactor `Template` so the template doesn't have to be created this way.
+      try template.template.createProject(
+        at: projectPath,
+        inSubdirectory: name != nil,
+        on: localFileSystem,
+        terminal: terminal
+      )
+    } catch let nsError as NSError {
```

Comment:
Plain `Error` type has the `localizedDescription` property too, is there a need to cast to `NSError` here after all?

Reply:
I tried to look for the implementation of `localizedDescription` for `Error`, and all I found is [this](https://github.com/apple/swift/blob/4fbdaf53eadea47e3dabe13f9111f5d60cab6bce/stdlib/public/Darwin/Foundation/NSError.swift#L213):

```swift
public extension Error {
  /// Retrieve the localized description for this error.
  var localizedDescription: String {
    return (self as NSError).localizedDescription
  }
}
```

I don't know if all errors can be cast as `NSError`. I suppose I can just do this:

```swift
catch let error {
  throw ProjectInitialisationError.cannotCreateWithTemplate(template, reason: error.localizedDescription)
}
```

but I don't know what would happen to errors that can't be cast as `NSError`.

I think if carton ever uses libSwiftPM directly, then we can use its diagnostics engine, and all these "catch and throw something different" won't be needed anymore.

Reply:
This option seems legit to me:
```swift
catch let error {
  throw ProjectInitialisationError.cannotCreateWithTemplate(template, reason: error.localizedDescription)
}
```

Reply:
I just tried this in the REPL:

```swift
enum ExampleError: Error, CustomStringConvertible { 
    case example1 
    var description: String { 
        "this is an example error" 
    }
}

let anError = ExampleError.example1
print(anError.localizedDescription)

error: repl.swift:9:15: error: value of type 'ExampleError' has no member 'localizedDescription'
print(anError.localizedDescription)
      ~~~~~~~ ^~~~~~~~~~~~~~~~~~~~
```

It gives the same error if `ExampleError` doesn't conform to `CustomStringConvertible`.

Reply:
Well, this `NSError` business gets ugly really quickly, consider this:

```swift
import Foundation

enum ExampleError: Error, CustomStringConvertible {
    case example1 
    var description: String { 
        "this is an example error" 
    }
}

do {
    throw ExampleError.example1
} catch let nsError as NSError {
    print("nserror: \(nsError.localizedDescription)")
} catch {
    print("error \(error)")
}
```

prints `nserror: The operation couldn’t be completed. (ConcurrencyPlayground.ExampleError error 0.)`. That is, any error can be cast to `NSError`, but its `localizedDescription` ignores the `description` value. The `catch let error` branch is never evaluated, because the first `NSError` branch always takes precedence no matter the actual type of the error. At least from this perspective there should be only one branch, either handling it as `NSError` or `Error`. Having two branch is redundant.